text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,514 @@
+
+Alternatively, users can use `cdk deploy --construct-upgrade` with construct upgrade option to validate
+their construct upgrade changes before deployment starts.
+
+In the initial release of the CDK construct upgrade tool, the CLI supports migration of VPC-related constructs
+from `aws-cdk-lib/aws-ec2` to `@aws-cdk/aws-ec2-alpha` and `Table` to `TableV2` in `aws-cdk-lib/aws-dynamodb`.
+While the tool is designed to handle a wide range of construct migrations, the migration guides for these two
+upgrades will be the first available, with additional guides planned for future releases.
+
+#### Migration Strategies
+
+Depending on the construct or module being migrated, the deployment step differs. The framework automatically
+selects one of two migration strategies based on the resources being migrated:
+
+1. In-place Migration - This strategy is suitable for constructs where the changes do not involve resource
+removal or introduction of new underlying resource types (e.g., constructs that are compatible with
+CloudFormation's stack refactoring). 
+
+Users are expected to refactor their current CDK stack to use V2 constructs, ensuring that the resource
+configuration matches by following the migration guide. Users must then specify a `refactor.json` file, which
+is needed for the CloudFormation stack refactoring feature to map logical ID changes from the old construct to
+the new one. This file helps ensure that CloudFormation can track the logical resource IDs during the
+migration process, avoiding issues with resource recreation.
+
+Finally, the migration proceeds by validations. If validations are successful, the framework continues to invoke
+the `aws cloudformation execute-stack-refactor --stack-refactor-id <id>` command. This command triggers the stack
+refactoring operation, which helps to retain the existing resources and applies the necessary updates to the
+infrastructure. The goal is to enable a smooth migration with minimal disruption to running services, avoiding
+downtime or resource replacements, and ensuring the new constructs function seamlessly with the existing
+CloudFormation stack.
+
+2. Retain-Remove-Import Migration - This approach is used when the underlying resource has significant changes,
+such as transitioning from a custom resource to a native L1 resource, as seen in the migration from `Table` to
+`TableV2`. For this type of migration, users need to ensure that `RemovalPolicy.RETAIN` is set on the V1 construct
+in the stack to prevent the deletion of the resource during the migration. The `RemovalPolicy.RETAIN` ensures that
+the existing resource is retained by CloudFormation, avoiding accidental deletion when the CDK stack is updated.
+
+If the retain policy is not already set, users must first deploy the retain policy change to ensure that
+CloudFormation will retain the resource during the migration process. It's a good practice to keep the stateful
```

Comment:
Totally something we could automate.

For example, it is easily possible to

- Download the current template
- Patch the current template, changing/adding `DeletionPolicy`s wherever necessary
- Deploy that template

At no point did the user have to change their code to change those `DeletionPolicy`s; we can do that for them.


-------

I think (2) is a different, less-safe form of (1) Stack Refactoring that we can offer, but it's an operation that slots into the same location in the workflow I think.